xla/service/gpu/runtime/nccl_collective_permute_thunk.cc

```diff
@@ -53,6 +53,34 @@ absl::StatusOr<const int64_t> GetCurrentId(
           : current_logical_id.computation_id;
   return current_id;
 }
+
+bool IsLocalPeerTransfer(
+    const NcclP2PConfig::SourceTargetMapEntry& source_target,
+    se::Stream& stream, const int64_t current_id, const int64_t device_count) {
+  const std::optional<int64_t> source_id = source_target.source;
+  const std::optional<int64_t> target_id = source_target.target;
+  // Mixing nccl p2p with p2p memcopy will cause random deadlocks, namely
+  // when calling nccl call and cuda memcpy p2p together(which both are
+  // synchronizing devices), in this case if this rank is sending across host
+  // using a nccl call but receiving from a local peer which is going through
+  // cuda api, the deadlock could happen because nccl cannot ensure the
+  // order of cuda api calls.
+  // We determine if it's a local peer by the following conditions:
+  // 1. Both source and target IDs are present and they are within a node
+  // 2. Source ID is present and is within the node, but target ID is not
+  // present.
+  // 3. Target ID is present and is within the node, but source ID is not
+  // present.
+  int64_t host_id = (current_id / device_count);
+  if (source_id && target_id) {
+    return (host_id == (*source_id / device_count)) &&
+           (host_id == (*target_id / device_count));
+  }
+  if (source_id && host_id != *source_id / device_count) return false;
+  if (target_id && host_id != *target_id / device_count) return false;
+  return true;
```

Comment:
Same comment as on the earlier PR. I think this can be simplified. You can check src and target ID separately, if present. Also the comment can be simpler in the same way. 
You really just care if you're sending to a device within the same hiost or receiving from one or both, no?

Reply:
forgot to remove the redundant check above. Updated the comment.